2020/dec16/rules/rule.go

```diff
@@ -0,0 +1,64 @@
+package rules
+
+// NewRule is an ugly way to create a rule
+func NewRule(min1, max1, min2, max2 int, isDeparture bool) (Rule, error) {
+	// todo validate min < max
+	return Rule{
+		ranges: []ruleRange{
+			{min1, max1},
+			{min2, max2},
+		},
+		isDeparture: isDeparture,
+		lock: -1,
+	}, nil
+}
+
+// Rule defines one rule
+type Rule struct {
+	ranges      []ruleRange
+	positions   int
+	maxMask     int
+	isDeparture bool
+	lock        int
+}
+
+// IsValid returns whether the value matches the rule
+func (ru *Rule) IsValid(v int) bool {
+	for _, rg := range ru.ranges {
+		if rg.isValid(v) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ru *Rule) MatchTicket(ticket Ticket) {
+	for i, v := range ticket {
+		if !ru.IsValid(v) {
+			ru.positions &= (1 << i) ^ ru.maxMask
+		}
+	}
+}
+
+func (ru *Rule) TryLock() (int, bool) {
+	if ru.lock >= 0 {
+		return ru.lock, false
+	}
+	// if there is only one position left
+	if ru.positions&(ru.positions-1) == 0 {
+		for ru.positions > 0 {
+			ru.lock ++
+			ru.positions >>= 1
+		}
+		return ru.lock, true
+	}
+	return 0, false
+}
+
+type ruleRange struct {
+	min, max int
+}
+
+func (rr *ruleRange) isValid(v int) bool {
```

Comment:
food for thought: should such funcs use `func (rr ruleRange) isValid(v int) bool` instead ?